ethcore/src/snapshot/service.rs

```diff
@@ -327,9 +328,74 @@ impl Service {
 
 	// replace one the client's database with our own.
 	fn replace_client_db(&self) -> Result<(), Error> {
-		let our_db = self.restoration_db();
+		let rest_db = self.restoration_db();
+
+		let cur_chain = &self.client.chain();
+		let next_db = self.restoration_db_handler.open(&rest_db)?;
+		let next_chain = BlockChain::new(Default::default(), &[], next_db.clone());
+		let next_chain_info = next_chain.chain_info();
+
+		// Try to include every block that will need to be downloaded from the current chain
+		// Break when no more blocks are available from it.
+		match (next_chain_info.ancient_block_number, next_chain_info.first_block_number) {
+			(Some(next_ancient_block), Some(next_first_block)) if next_ancient_block + 1 < next_first_block => {
+				trace!(target: "snapshot", "Trying to import ancient blocks from {} to {}",
+					next_ancient_block, next_first_block,
+				);
+
+				let mut block_number = next_ancient_block + 1;
+				let mut batch = DBTransaction::new();
+				let mut parent_diff = None;
+
+				while block_number < next_first_block {
+					let chain = cur_chain.read();
+
+					if let Some(block_hash) = chain.block_hash(block_number) {
+						let block = chain.block(&block_hash);
+						let block_receipts = chain.block_receipts(&block_hash);
+
+						match (block, block_receipts) {
+							(Some(block), Some(block_receipts)) => {
+								let diff = block.difficulty();
+								let raw_block = block.into_inner();
+								let block_receipts = block_receipts.receipts;
+
+								next_chain.insert_unordered_block(&mut batch, &raw_block, block_receipts, parent_diff, false, true);
+								parent_diff = Some(diff);
+							},
+							_ => (),
+						}
+					// Break if we already imported some blocks in the current batch and there
+					// are no more left
+					} else if parent_diff.is_some() {
```

Comment:
why only break if `parent_diff` is `Some`? if there are no more blocks we can import it seems like it should _always_ break the loop.

Reply:
Yep that's right, didn't push this change yet.